### PR TITLE
Add tests to cart

### PR DESCRIPTION
**Pull request by maciekjablonski**

### CODE REVIEW

**Comment by artursmet** [comment]
@maciekjablonski Please fix the build. Also some description will be nice, when there's no ticket attached to a pull request.

**Comment by maciekjablonski** [comment]
Refs #727

File: saleor/cart/models.py [hunk 1]
```diff
@@ -161,6 +162,8 @@ def get_total(self, **kwargs):
                      for item in self.lines.all()]
         if not subtotals:
             raise AttributeError('Calling get_total() on an empty item set')
+        # second sum argument is default 0, so sum(subtotals) raise
+        # TypeError: unsupported operand type(s) for +: 'int' and 'Price'
```
Comment (krzysztofwolski) [inline saleor/cart/models.py]: zero = Price(0, currency=settings.DEFAULT_CURRENCY)
total = sum(list, zero)

File: tests/test_cart.py [hunk 2]
```diff
@@ -450,7 +458,10 @@ def test_view_cart(client, product_in_stock, request_cart):
     assert response.status_code == 200
 
 
-def test_view_update_cart_quantity(client, product_in_stock, request_cart):
+def test_view_update_cart_quantity(client, product_in_stock, monkeypatch, request_cart):
```
Comment (patrys) [inline tests/test_cart.py]: Please keep fixtures sorted.

File: tests/test_cart.py [hunk 3]
```diff
@@ -450,7 +458,10 @@ def test_view_cart(client, product_in_stock, request_cart):
...
+    def side_effect(price, currency):
+        return price
+    monkeypatch.setattr('saleor.cart.views.to_local_currency', Mock(side_effect=side_effect))
```
Comment (patrys) [inline tests/test_cart.py]: This looks like a candidate for a fixture.

File: tests/test_cart.py [hunk 4]
```diff
@@ -524,3 +535,171 @@ def test_cart_summary_page_empty_cart(client, request_cart):
...
+    group = ProductGroup([Mock(is_shipping_required=Mock(return_value=False)),
+                          Mock(is_shipping_required=Mock(return_value=False))])
+    assert not group.is_shipping_required()
+    group = ProductGroup([Mock(is_shipping_required=Mock(return_value=True)),
+                          Mock(is_shipping_required=Mock(return_value=False))])
+    assert group.is_shipping_required()
+
+
+def test_cart_queryset(customer_user):
+    cart1 = Cart.objects.create(status=Cart.SAVED)
```
Comment (patrys) [inline tests/test_cart.py]: Can we give these meaningful names? What sets `cart1` and `cart2` apart?

File: tests/test_cart.py [hunk 5]
```diff
@@ -524,3 +535,171 @@ def test_cart_summary_page_empty_cart(client, request_cart):
...
+    cart2 = Cart.objects.create(user=customer_user)
+
+    anonymous = Cart.objects.anonymous()
```
Comment (patrys) [inline tests/test_cart.py]: `anonymous_carts`?

File: tests/test_cart.py [hunk 6]
```diff
@@ -524,3 +535,171 @@ def test_cart_summary_page_empty_cart(client, request_cart):
...
+
+    assert len(anonymous) == 1
+    assert anonymous[0] == cart1
+
+    saved = Cart.objects.saved()
+
+    assert len(saved) == 1
+    assert saved[0] == cart1
+
+    cart3 = Cart.objects.create(status=Cart.WAITING_FOR_PAYMENT)
```
Comment (patrys) [inline tests/test_cart.py]: See above for naming.

File: tests/test_cart.py [hunk 7]
```diff
@@ -524,3 +535,171 @@ def test_cart_summary_page_empty_cart(client, request_cart):
...
+
+    cart2.change_user(customer_user)
+
+    old_cart = Cart.objects.get(pk=cart1.pk)
+
+    assert old_cart.user == customer_user
+    assert old_cart.status == Cart.CANCELED
+
+
+def test_cart_line_str__repr__(product_in_stock, request_cart_with_item):
```
Comment (patrys) [inline tests/test_cart.py]: `repr` would be enough, use `repr()` instead of calling the `__repr__` magic method directly.

File: tests/test_cart.py [hunk 8]
```diff
@@ -524,3 +535,171 @@ def test_cart_summary_page_empty_cart(client, request_cart):
...
+    line.__setstate__((variant, 2, line.data))
+
+    assert line.quantity == 2
+
+
+def test_get_category_variants_and_prices(default_category, product_in_stock,
+                                          request_cart_with_item):
+    variant = product_in_stock.variants.get()
+    top_category = Category.objects.create(name='foo', slug='bar')
+    alt_category = Category.objects.create(name='bar', slug='foo')
```
Comment (patrys) [inline tests/test_cart.py]: Why is this needed?

File: tests/test_cart.py [hunk 9]
```diff
@@ -524,3 +535,171 @@ def test_cart_summary_page_empty_cart(client, request_cart):
...
+def test_update_view_must_be_ajax(customer_user, rf):
+    request = rf.post('/')
+    request.user = customer_user
+    result = update(request, 1)
+    assert result.status_code == 302
+
+
+def test_get_or_create_db_cart(customer_user, db, rf):
+
+
```
Comment (patrys) [inline tests/test_cart.py]: What happened here?

File: tests/test_cart.py [hunk 10]
```diff
@@ -524,3 +535,171 @@ def test_cart_summary_page_empty_cart(client, request_cart):
...
+    def view(request, cart, *args, **kwargs):
```
Comment (patrys) [inline tests/test_cart.py]: Why not pass a callable mock? Mock has a `assert_called_once_with` method.

File: tests/test_cart.py [hunk 11]
```diff
@@ -524,3 +535,171 @@ def test_cart_summary_page_empty_cart(client, request_cart):
...
+        return HttpResponse()
+
+    v = utils.get_or_create_db_cart()(view)
```
Comment (patrys) [inline tests/test_cart.py]: `v` is a bad name.

**Comment by artursmet** [comment]
@patrys Could you review?

**Comment by patrys** [comment]
The comment about `__repr__` still holds. I'm also not a fan of query length assertions, these will break if you create any objects in your migrations.

File: tests/test_cart.py [hunk 12]
```diff
@@ -549,3 +562,168 @@ def test_total_with_discount(client, sale, request_cart, product_in_stock):
...
+
+
+def test_cart_queryset(customer_user):
+    saved_cart = Cart.objects.create(status=Cart.SAVED)
+    customer_cart = Cart.objects.create(user=customer_user)
+
+    anonymous_cart = Cart.objects.anonymous()
+
+    assert len(anonymous_cart) == 1
+    assert anonymous_cart[0] == saved_cart
```
Comment (patrys) [inline tests/test_cart.py]: Can we change the above two to `assert anonymous_carts.filter(pk=saved_cart.pk).exists()`?

File: tests/test_cart.py [hunk 13]
```diff
@@ -549,3 +562,168 @@ def test_total_with_discount(client, sale, request_cart, product_in_stock):
...
+    group = ProductGroup([Mock(is_shipping_required=Mock(return_value=True)),
+                          Mock(is_shipping_required=Mock(return_value=False))])
+    assert group.is_shipping_required()
+
+
+def test_cart_queryset(customer_user):
+    saved_cart = Cart.objects.create(status=Cart.SAVED)
+    customer_cart = Cart.objects.create(user=customer_user)
+
+    anonymous_cart = Cart.objects.anonymous()
```
Comment (patrys) [inline tests/test_cart.py]: This should be plural. It's a collection (queryset).

File: tests/test_cart.py [hunk 14]
```diff
@@ -54,6 +58,11 @@ def cancelled_user_cart(customer_user):
     return Cart.objects.get_or_create(user=customer_user,
                                       status=Cart.CANCELED)[0]
 
+@pytest.fixture()
+def local_currency(monkeypatch):
+    def side_effect(price, currency):
+        return price
+    monkeypatch.setattr('saleor.cart.views.to_local_currency', Mock(side_effect=side_effect))
```
Comment (maciekjablonski) [inline tests/test_cart.py]: fixed

File: tests/test_cart.py [hunk 15]
```diff
@@ -2,6 +2,8 @@
 
 import json
 from uuid import uuid4
+from django.core.urlresolvers import reverse
+from django.http import HttpResponse
```
Comment (maciekjablonski) [inline tests/test_cart.py]: fixed

**Comment by maciekjablonski** [comment]
@patrys and @krzysztofwolski review please

File: saleor/cart/models.py [hunk 16]
```diff
@@ -47,7 +48,8 @@ def canceled(self):
         return self.filter(status=CartStatus.CANCELED)
 
     def save(self):
-        self.update(status=CartStatus.SAVED)
+        self.exclude(status=CartStatus.SAVED).update(status=CartStatus.SAVED,
+                                               last_status_change=now())
```
Comment (krzysztofwolski) [inline saleor/cart/models.py]: Wrong argument intendent

File: saleor/cart/models.py [hunk 17]
```diff
@@ -47,7 +48,8 @@ def canceled(self):
         return self.filter(status=CartStatus.CANCELED)
 
     def save(self):
-        self.update(status=CartStatus.SAVED)
+        self.exclude(status=CartStatus.SAVED).update(status=CartStatus.SAVED,
+                                                     last_status_change=now())
```
Comment (patrys) [inline saleor/cart/models.py]: What does this do? When is it used?

File: saleor/cart/models.py [hunk 18]
```diff
@@ -46,9 +47,6 @@ def checkout(self):
     def canceled(self):
         return self.filter(status=CartStatus.CANCELED)
 
-    def save(self):
-        self.update(status=CartStatus.SAVED)
```
Comment (bogdal) [inline saleor/cart/models.py]: Could you explain why you removed this? Is it related to tests?

Comment (patrys) [inline saleor/cart/models.py]: I asked for dead code to be removed rather than tested.

Comment (maciekjablonski) [inline saleor/cart/models.py]: @bogdal as patrys said
